pyrealm/demography/canopy.py

```diff
@@ -0,0 +1,133 @@
+
+        self.crown_area_per_layer: float = community.cell_area * (
+            1 - canopy_gap_fraction
+        )
+        """Total crown area permitted in a single canopy layer, given the available
+        cell area of the community and its canopy gap fraction."""
+
+        self.n_layers: int = int(
+            np.ceil(self.total_community_crown_area / self.crown_area_per_layer)
+        )
+        """Total number of canopy layers."""
+
+        self.n_cohorts: int = community.number_of_cohorts
+        """Total number of cohorts in the canopy."""
+
+        # Find the closure heights of the canopy layers under the perfect plasticity
+        # approximation by solving Ac(z) - L_n = 0 across the community where L is the
+        # total cumulative crown area in layer n and above, discounted by the canopy gap
+        # fraction.
+
+        self.layer_heights: NDArray[np.float32] = np.zeros(
+            (self.n_layers, 1), dtype=np.float32
+        )
+        """Column vector of the heights of canopy layers."""
+
+        # Loop over the layers except for the final layer, which will be the partial
+        # remaining vegetation below the last closed layer.
+        starting_guess = self.max_stem_height
+        for layer in np.arange(self.n_layers - 1):
+            target_area = (layer + 1) * community.cell_area * (1 - canopy_gap_fraction)
+
+            # TODO - the solution here is predictably closer to the upper bracket, might
+            # be a better algorithm to find the root.
+            solution = root_scalar(
+                solve_community_projected_canopy_area,
+                args=(
+                    community.cohort_data["stem_height"].to_numpy(),
+                    community.cohort_data["crown_area"].to_numpy(),
+                    community.cohort_data["m"].to_numpy(),
+                    community.cohort_data["n"].to_numpy(),
+                    community.cohort_data["q_m"].to_numpy(),
+                    community.cohort_data["canopy_z_max"].to_numpy(),
+                    community.cohort_data["n_individuals"].to_numpy(),
+                    target_area,
+                    False,  # validate
+                ),
+                bracket=(0, starting_guess),
+                xtol=layer_tolerance,
+            )
+
+            if not solution.converged:
+                raise RuntimeError(
+                    "Estimation of canopy layer closure heights failed to converge."
+                )
+
+            self.layer_heights[layer] = starting_guess = solution.root
+
+        # Find relative canopy radius at the layer heights
+        self.stem_relative_radius: NDArray[np.float32] = (
+            calculate_relative_canopy_radius_at_z(
+                z=self.layer_heights,
+                stem_height=community.cohort_data["stem_height"].to_numpy(),
+                m=community.cohort_data["m"].to_numpy(),
+                n=community.cohort_data["n"].to_numpy(),
+                validate=False,
```

Comment:
`validate` is set to `False` in these function calls. Should there be an option to set this to true via this class?

Reply:
I should doc this better. The validation is intended to help people use the standalone functions correctly, but here the `Canopy.__init__` should guarantee the correct inputs, and so the validation is turned off to improve run time by avoiding repeated validation within the root solver call.

If it's broken, that's a developer issue with the `__init__` code, rather than something where a user could usefully turn validation back on.

Reply:
Comments added to code to explain this.